Support TS moduleResolution node16/nodenext by adding "types" to package.json's exports map

### On which framework/platform would you like to see this feature implemented?

React

### Which UI component is this feature-request for?

Other

### Please describe your feature-request in detail.

I'd like to be able to use `@aws-amplify/ui-react` within a TypeScript project with a tsconfig.json that has a `moduleResolution` value of `node16` or `nodenext`. When I currently do this, I get errors like:

```bash
Try `npm i --save-dev @types/aws-amplify__ui-react` if it exists or add a new declaration (.d.ts) file containing `declare module '@aws-amplify/ui-react';`
src/components/AccountMenu.tsx(2,51): error TS7016: Could not find a declaration file for module '@aws-amplify/ui-react'. '/Users/stickb/Code/niwc-pac/cosmos/node_modules/.pnpm/@aws-amplify+ui-react@3.5.7_bhvjxgt3qlsa6h77uyvrixez5u/node_modules/@aws-amplify/ui-react/dist/esm/index.js' implicitly has an 'any' type.
```

### Please describe a solution you'd like.

In order to fix this, Amplify UI libraries should include a `"types"` key within their package.json's exports maps. `@aws-amplify/ui-react` example:
Current:
```json
"exports": {
    ".": {
      "import": "./dist/esm/index.js",
      "require": "./dist/index.js"
    },
    ...
}
```
Requested:
```json
"exports": {
    ".": {
      "types": "./dist/types/index.d.ts",
      "import": "./dist/esm/index.js",
      "require": "./dist/index.js"
    },
    ...
}
```
Example of another library doing the same thing: https://github.com/tj/commander.js/pull/1704
TypeScript background: https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing

Note, as stated in TS docs, the `"types"` key must occur first

### We love contributors! Is this something you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request.
- [ ] ⚠️ This feature might incur a breaking change.

--- a/.changeset/gold-kiwis-mate.md
+++ b/.changeset/gold-kiwis-mate.md
@@ -1,0 +1,6 @@
+---
+"@aws-amplify/ui-react": patch
+"@aws-amplify/ui": patch
+---
+
+chore(react): adding types to exports in package json to support TS moduleResolution for node16/nodenext

--- a/packages/react/package.json
+++ b/packages/react/package.json
@@ -5,6 +5,7 @@
   "module": "dist/esm/index.js",
   "exports": {
     ".": {
+      "types": "./dist/types/index.d.ts",
       "import": "./dist/esm/index.js",
       "require": "./dist/index.js"
     },

--- a/packages/ui/package.json
+++ b/packages/ui/package.json
@@ -5,6 +5,7 @@
   "module": "dist/esm/index.js",
   "exports": {
     ".": {
+      "types": "./dist/types/index.d.ts",
       "import": "./dist/esm/index.js",
       "require": "./dist/index.js"
     },

